Add Config Sections

Hydrox6 opened the pull request:
Closes #11602

Certain configs have always been begging for some form of section system. When merging multiple plugins together, being able to separate their options would be very useful.

Sections can be toggled by clicking anywhere on them, and are persisted through the config. It is possible to nest configs inside one another. Each level of nesting adds a small but noticeable margin on the left side.

I have added sections to Chat Color (which desperately needed it) to show users that there are 2 sets of options, and to Ground Items to hide potentially gigantic item lists.

Feedback on the design of the sections is very much appreciated

![](https://giant.gfycat.com/FlakyFittingBumblebee.gif)

(Above gif is outdated, as the text is now RL orange. See comments below)

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/config/ConfigDescriptor.java:
```diff
@@ -24,26 +24,22 @@
 
 public class ConfigDescriptor
 {
+	@Getter
```
If all fields have `@Getter`, you can annotate the class with it instead for brevity.

raiyni commented on runelite-client/src/main/java/net/runelite/client/config/ConfigSection.java:
```diff
@@ -0,0 +1,47 @@
+
+	String description();
+
+	String warning() default "";
```
This looks unused?

Hydrox6 commented:
Updated the visuals from feedback in the Discord
![image](https://user-images.githubusercontent.com/2979691/62429247-524c6000-b704-11e9-83c3-75b8403b1dfb.png)

adwitkow commented:
this will probably close #6707

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -127,6 +138,13 @@
+		final BufferedImage sectionRetractIcon = ImageUtil.getResourceStreamFromClass(ConfigPanel.class, "config_section_retract_icon.png");
+		SECTION_RETRACT_ICON = new ImageIcon(sectionRetractIcon);
+		SECTION_RETRACT_ICON_HOVER = new ImageIcon(ImageUtil.alphaOffset(sectionRetractIcon, -100));
+		final BufferedImage sectionExpandIcon = ImageUtil.rotateImage(sectionRetractIcon, Math.PI);
```
This should probably be `ImageUtil.flipImage()`?
```suggestion
		final BufferedImage sectionExpandIcon = ImageUtil.flipImage(sectionRetractIcon, false, true);
```

Hydrox6 commented:
Updated with @Nightfirecat's suggestions. The icon is now slightly smaller, but it looks better in my opinion.

Hydrox6 commented:
Alright, I've added sections for all the plugins currently in the client, and also split up the hardwood grove swap since it fell into two headings.

Nightfirecat commented:
04cbc88 and a3e3aed can be merged imo. Adding config sections support and adding a saving mechanism for it should be one change.

Nightfirecat commented:
In some config sections, expanding a config section can cause the section header underline to narrow, and the arrow to move left with it. For example, in the MES config panel, collapse all sections then toggle the "Item Swaps" section to see the following behavior:

![config-sections-narrowing](https://user-images.githubusercontent.com/2199511/82524554-c8e50d80-9b1e-11ea-8145-6c533180f127.png)

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerConfig.java:
```diff
@@ -28,14 +28,24 @@
 public interface LootTrackerConfig extends Config
 {
+	@ConfigSection(
+		name = "Ignored",
```
```suggestion
		name = "Ignored Entries",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/implings/ImplingsConfig.java:
```diff
@@ -43,11 +44,20 @@
 	}
 
+	@ConfigSection(
+		name = "Impling Configuration",
```
Maybe something more like this? (maybe "Impling Tier" makes more sense? idk)
```suggestion
		name = "Impling Type Settings",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java:
```diff
@@ -36,11 +36,40 @@
 	String GROUP_NAME = "runelite";
 
+	@ConfigSection(
+		name = "Window Settings",
```
The word "Settings" is used here, but "Configuration" and "Options" are used in other configs. Of these choices I prefer "Settings" personally, but would like for one word or the other to be used in all cases rather than have a mix of them.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java:
```diff
@@ -29,10 +29,27 @@
 {
+	@ConfigSection(
+		name = "Charges Configuration",
+		description = "Configuration for all charges that can be shown",
```
```suggestion
		description = "Configuration for which charges should be displayed",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java:
```diff
@@ -28,15 +28,25 @@
 {
+	@ConfigSection(
+		name = "Highlight Options",
+		description = "",
```
```suggestion
		description = "Toggle highlighted players by type (self, friends, etc.) and choose their highlight colors",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/randomevents/RandomEventConfig.java:
```diff
@@ -28,10 +28,19 @@
 public interface RandomEventConfig extends Config
 {
+	@ConfigSection(
+		name = "Random Events",
```
```suggestion
		name = "Notification Settings",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/randomevents/RandomEventConfig.java:
```diff
@@ -28,10 +28,19 @@
 {
+	@ConfigSection(
+		name = "Random Events",
+		description = "",
```
```suggestion
		description = "Choose which random events will trigger notifications when spawned",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/runecraft/RunecraftConfig.java:
```diff
@@ -28,26 +28,49 @@
 {
+	@ConfigSection(
+		name = "Rift Configuration",
+		description = "All the options that control Rifts",
```
```suggestion
		description = "Abyss rift overlay settings",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerConfig.java:
```diff
@@ -28,11 +28,20 @@
+	@ConfigSection(
+		position = 0,
+		name = "Overlay",
+		description = "Options that relate to the overlay",
```
```suggestion
		description = "Canvas overlay options",
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerConfig.java:
```diff
@@ -115,7 +127,8 @@ default XpPanelLabel onScreenDisplayModeBottom()
 		name = "Top-left XP info label",
-		description = "Configures the information displayed in the top-left of XP info box"
+		description = "Configures the information displayed in the top-left of XP info box",
+		section = overlaySection
```
The last four options (`xpPanelLabel[1234]`) marked here **are not part of the overlay**. These are the labels of the skill boxes displayed in the side panel.

Hydrox6 commented on runelite-client/src/main/java/net/runelite/client/config/ConfigManager.java:
```diff
@@ -468,8 +468,31 @@ public ConfigDescriptor getConfigDescriptor(Config configurationProxy)
+						m.getDeclaredAnnotation(ConfigSection.class)
+					);
+				}
+				catch (IllegalAccessException e)
+				{
+					return null;
+				}
```
Due to having the keyName be the contents of the field that the annotation is attached to, I have to reflect to get it. If I don't have this try/catch it won't compile since it can throw, and if it does throw then it's not something we'd need. If you have a cleaner way to do this I would definitely love to know.

Nightfirecat replied:
I see, that's unfortunate. I maintain we should log for failures here, even if we don't expect them. I'd rather log some warnings than have silent failures in the event something went wrong.

Hydrox6 commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -205,11 +229,95 @@ void init(PluginConfigurationDescriptor pluginConfig)
+			sectionContents.setVisible(isOpen);
+
+			final JToggleButton sectionToggle = new JToggleButton(isOpen ? SECTION_RETRACT_ICON : SECTION_EXPAND_ICON);
+			sectionToggle.setRolloverIcon(isOpen ? SECTION_RETRACT_ICON_HOVER : SECTION_EXPAND_ICON_HOVER);
```
As discussed on Discord, this seems to be fine as all other buttons in the client do it, and unless you know it's there it's barely noticeable.

Hydrox6 commented:
Fixed [this](https://github.com/runelite/runelite/pull/9572#issuecomment-631878809) weird issue, added a log message if the reflection fails, and updated to work with the new Ignored Events configitem

Adam- commented on runelite-client/src/main/java/net/runelite/client/config/ConfigSection.java:
```diff
@@ -0,0 +1,45 @@
+
+	boolean closedByDefault() default false;
+
+	boolean displayAfterItems() default false;
```
I think this can be removed and replace with section priority relative to sectionless item priority.

abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -205,11 +230,92 @@ void init(PluginConfigurationDescriptor pluginConfig)
+			sectionName.setToolTipText("<html>" + name + ":<br>" + cs.description() + "</html>");
+			sectionHeader.add(sectionName, BorderLayout.WEST);
+
+			final JToggleButton sectionToggle = new JToggleButton(isOpen ? SECTION_RETRACT_ICON : SECTION_EXPAND_ICON);
```
why are you using a toggle button, that doesn't have a toggled icon, that emulates toggling via some other method?

Hydrox6 replied:
It used to be an IconButton (I think that's what it was called) before that got removed, and when I looked at what replaced it, this was what I found

abextm replied:
You should either a) use the toggle functionality or b) not use a toggle button

abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -205,11 +230,92 @@ void init(PluginConfigurationDescriptor pluginConfig)
+					toggleSection(cd, sectionToggle, sectionContents, csd.getKey());
+				}
+			};
+			sectionToggle.addMouseListener(adapter);
```
this should be an actionlistener

abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListPanel.java:
```diff
@@ -101,6 +104,7 @@
 	private final JScrollPane scrollPane;
 	private final FixedWidthPanel mainPanel;
 	private List<PluginListItem> pluginList;
+	private final Table<String, String, Boolean> sectionExpandStates = HashBasedTable.create();
```
Why can't this just be a `Map<ConfigSectionDescriptor, Boolean>`?
Also this should probably go in plugin, it has no relation to the state of the plugin list

Hydrox6 replied:
You did suggest `PluginListPanel` earlier, but yeah I get that.
Should I make it all static in ConfigPlugin or inject ConfigPlugin into ConfigPanel?

abextm replied:
Inject it

Hydrox6 replied:
Injecting it causes
```
java.lang.IllegalStateException: Recursive load of: net.runelite.client.plugins.config.PluginListPanel.<init>()```
